File: src/Lex.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Lex {

    private ArrayList<String> reserved;
    private ArrayList<Token> tokens;

    private int linePos;
    private int charPos;

    public Lex() {

        this.reserved = new ArrayList<>();
        this.tokens = new ArrayList<>();

        reserved.add("principal");
        reserved.add("entero");
        reserved.add("real");
        reserved.add("logico");
        reserved.add("si");
        reserved.add("mientras");
        reserved.add("regresa");
        reserved.add("verdadero");
        reserved.add("falso");


    }

    private boolean isEspecialChar(char val) {
        switch (val) {
            case '\t':
            case '\n':
            case ' ':
                return true;
            default:
                return false;
        }

    }

    private boolean isReservedWord(String word) {
        if(this.reserved.contains(word)) {
            this.addToken(Token.KEYWORD, word);
        }
        return this.reserved.contains(word);
    }

    private boolean isOperatorsChar(Character val, String line, int pos) {
        switch (val) {
            case '+':
            case '-':
            case '*':
            case '/':
            case '^':
                this.addToken(Token.ARITMETHIC, val.toString());
                return true;
            case '<':
            case '>':
                this.addToken(Token.RELATIONAL, val.toString());
                return true;
            case '=':
                boolean isEqualOperator = this.validEqualOperator(Token.ASSIGMENT, pos, line);
                if(!isEqualOperator) {
                    this.addToken(Token.ASSIGMENT, val.toString());
                }
                return true;
            default:
                return false;

        }
    }

    private boolean validEqualOperator(int tokenType, int characterPos, String line) {
        int position = characterPos;
        if(Token.ASSIGMENT == tokenType) {
            String str = "";
            char character = line.charAt(position);
            str += character;
            position++;
            character = line.charAt(position);
            str += character;
            if(str.equals("==")) {
            this.addToken(Token.RELATIONAL, str);
                this.charPos++;
                return true;
            }
            return false;
        }
        return false;
    }

    private boolean isPunctuationChar(Character val) {
        switch (val) {
            case ',':
            case ';':
            case '(':
            case ')':
            case '{':
            case '}':
                this.addToken(Token.PUNCTUATION, val.toString());
                return true;
            default:
                return false;
        }
    }

    private boolean isNumberChar(String val) {

        String str = "";
        int dots = 0;
        for(int i=0; i<val.length(); i++) {
            Character character = val.charAt(i);
            if(character.isDigit(character)) {
                str += character;
                continue;
            }
            if(this.validDot(character)) {
                dots ++;
                str += character;
                if(dots > 1) {
                    this.addToken(Token.ERROR, str);
                    return false;
                }
                str += character;
            }

        }
        if(dots == 1){
            this.addToken(Token.REAL, str);
            return true;
        } else {
            this.addToken(Token.ENTERO, str);
            return true;
        }
    }

    private boolean validDot(char val) {
        return '.' == val;
    }

    private void addToken(int tokenType, String val) {
        this.tokens.add(new Token(tokenType, val, this.linePos, this.charPos));
    }

    private boolean isIdentifier(String val) {

        String id = "";
        for(int i=0; i<val.length(); i++) {
            Character character = val.charAt(i);
            if(character.isLetter(i) && i==0) {
                id += character;
            } else {
                this.addToken(Token.ERROR, id);
                return false;
            }
        }
        this.addToken(Token.IDENTIFIER, id);
        return true;

    }


    private void states(String currentLine) {

        String str = "";
        for (int i = 0; i < currentLine.length(); i++) {
            char currentChar = currentLine.charAt(i);
            if(!this.isEspecialChar(currentChar)) {
                if(this.isPunctuationChar(currentChar)) {
                    str += currentChar;
                    this.charPos++;
                    continue;
                }  else {
                    str += currentChar;
                    if(this.isOperatorsChar(currentChar, currentLine, i)) {
                        i = this.charPos;
                        continue;
                    } else {
                        if(this.isReservedWord(str)) {
                            continue;
                        } else {
                            if(this.isNumberChar(str)) {
                                continue;
                            }
                            if(this.isIdentifier(str)) {
                                continue;
                            }
                        }
                    }
                }
            }

        }
    }

    public void analyze(String filename) {

        try {

            FileReader fr = new FileReader(filename);
            BufferedReader br = new BufferedReader(fr);

            String sCurrentLine;

            charPos = 0;
            linePos = 0;

            while ((sCurrentLine = br.readLine()) != null) {
                this.states(sCurrentLine);
                this.linePos++;
                System.out.println(sCurrentLine);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: lex.py
import token as Token
tokens = []
errors = []
symbols = {}
keywords = ["entero", "real", "logico", "regresa", "si", "mientras", "principal", "verdadero", "falso"]
initLower = ord('a')
finalLower = ord('z')
minNum = ord('0')
maxNum = ord('9')
initUpper = ord('A')
finalUpper = ord('Z')

def validateToken(typeToken, valToken, line, char):
  isIdentifier = False
  if isSpecialChar(valToken):
    return -1
  if typeToken == Token.ERROR:
    errors.append(Token.build(typeToken, valToken, line, char))
    return -1
  if typeToken == Token.IDENTIFIER:
    isIdentifier = True
    if (valToken in keywords):
      typeToken = Token.KEYWORD
      addToken(typeToken, valToken, line, char)
      return
    else:
      temp = validate_types(valToken)
      if temp != -1:
        print(valToken, Token.types.get(temp))
        typeToken = temp
        isIdentifier = False
  if(isIdentifier) :
    addToken(typeToken, valToken, line, char)
    if valToken not in symbols:
      symbols[valToken] = Token.build(typeToken, valToken, line, char)
  else :
    addToken(validate_types(valToken), valToken, line, char)
  
def addToken(tokenType, valToken, line, char):
  if(tokenType >= 0 and tokenType <= 9):
    tokens.append(Token.build(tokenType, valToken, line, char))

def validate_types(val) :
  if (val == '+' or val == '-' or val == '*' or val == '/' or val == '^'):
    return Token.ARITMETHIC
  elif (val == '<' or val == '>'):
    return Token.RELATIONAL
  elif (val == ';' or val == ',' or val == '(' or val == ')' or val == '{' or val == '}'):
    return Token.PUNCTUATION
  elif (val == '!' or val == '&' or val == '|'):
    return Token.LOGICAL
  elif (val == '=') :
    return Token.ASSIGMENT
  elif (val == '==') :
    return Token.RELATIONAL
  else:
    return -1

def isSpecialChar(val) :
  if val == '\n':
    return True
  elif val == '\t': 
    return True
  elif val == '\r': 
    return True
  elif val == ' ':
    return True
  
  return False

def isUnaryChar(val) :
  unaryChars = {
    ';':True,
    ',':True,
    '(':True,
    ')':True,
    ')':True,
    '{':True,
    '}':True,
    '!':True,
    '&':True,
    '|':True,
    '<':True,
    '>':True,
    '+':True,
    '-':True,
    '*':True,
    '/':True,
    '^':True
  }
  return unaryChars.get(val, False)
    
def canBeLowerLetter(val) :
  initLimit = ord('a')
  finalLimit = ord('z')
  return (ord(val) >= initLimit and ord(val) <= finalLimit)

def canBeNumber(val) :
  numChar = ord(val)
  minNum = ord('0')
  maxNum = ord('9')
  isInRangeNumber = (minNum <= numChar and maxNum >= numChar)
  return isInRangeNumber

def canBeIdentifier(val) :
  numChar = ord(val)
  isInRangeLowerCase = (initLower <= numChar and finalLower >= numChar)
  isInRangeUpperCase = (initUpper <= numChar and finalUpper >= numChar)
  isInRangeNumber = (minNum <= numChar and maxNum >= numChar)

  return (isInRangeLowerCase or isInRangeUpperCase or isInRangeNumber)

def analyze(filename):
  with open(filename, 'r') as file:
    NO_TOKEN_TYPE = -1
    line = file.readline()
    linePos = 1
    charPos = 0
    tokenOption = NO_TOKEN_TYPE
    charInit = 0
    charVal = ''

    while line:
      charPos = 0
      while charPos < len(line):
        character = line[charPos]

        if isSpecialChar(character) :
          validateToken(tokenOption, charVal, linePos, charInit)
          tokenOption = NO_TOKEN_TYPE
          charPos += 1
          charVal = ""
          continue
        
        if isUnaryChar(character) :
          tokenType = validate_types(character)
          validateToken(tokenOption, charVal, linePos, charInit)
          validateToken(tokenType, character, linePos, charPos)
          tokenOption = NO_TOKEN_TYPE
          charPos += 1
          charVal = ""
          continue

        if tokenOption == NO_TOKEN_TYPE and canBeLowerLetter(character) :
          charInit = charPos
          charVal += character
          charPos += 1
          tokenOption = Token.IDENTIFIER
          continue

        if tokenOption == Token.IDENTIFIER and canBeIdentifier(character) :
          charVal += character
          charPos += 1
          continue

        if tokenOption == NO_TOKEN_TYPE and canBeNumber(character) :
          charInit = charPos
          charVal += character
          charPos += 1
          tokenOption = Token.INTEGER
          continue

        if (tokenOption == Token.INTEGER or tokenOption == Token.REAL) and canBeNumber(character) :
          charVal += character
          charPos += 1
          continue

        if tokenOption == Token.INTEGER and character == '.' :
          charVal += character
          charPos += 1
          character = line[charPos]
          print("canBeNumer", character)
          if canBeNumber(character) :
            charVal += character
            charPos += 1
            tokenOption = Token.REAL
            continue
          else :
            charVal += character
            charPos += 1
            tokenOption = Token.ERROR
            continue

        if character == '=' :
          validateToken(tokenOption, charVal, linePos, charPos)   
          charVal = ""
          tokenOption = NO_TOKEN_TYPE
          nextChar = line[charPos+1]
          if nextChar == '=' :
            validateToken(Token.RELATIONAL, '==', linePos, charPos)
            charPos += 2
            continue
          else :
            validateToken(Token.ASSIGMENT, '=', linePos, charPos)
            charPos += 1
            continue

        tokenOption = Token.ERROR
        charPos += 1
        charVal += character
      
      line = file.readline()
      linePos += 1
    validateToken(tokenOption, charVal, linePos-1, charInit)
    return tokens, errors, symbols

        

      
